### PR TITLE
Stop vendoring noVNC

### PR DESCRIPTION
**Pull request by yuvipanda**
We have been vendoring noVNC to get vnc_lite.html, maintaining a patch file so we could upgrade the version of noVNC used if needed.

noVNC publishes a JS library [onnpm](https://www.npmjs.com/package/@novnc/novnc)
that we can easily use instead, and stop vendoring the whole package! This brings us the following massive advantages:

1. No more vendoring an entire package in!
2. Upgrades of noVNC client become practically trivial
3. We can change the frontend as we wish much more easily, compared to having to maintain a patch file as we do now. PRs like https://github.com/jupyterhub/jupyter-remote-desktop-proxy/pull/7 become easier.
4. There's also some possible licensing issues around us vendoring this anyway https://github.com/jupyterhub/jupyter-remote-desktop-proxy/issues/27#issuecomment-1737925373, this will get rid of that

This PR:

- Adds a `package.json` to include the noVNC js package
- Adds a `webpack.config.js` to provide JS bundling, so we can just ship a single `viewer.js` file that has everything.
- Adds appropriate bits (stolen and adapted from jupyterhub's config) to package the built JS in the final wheel as necessary.
- Upgrade from noVNC 1.2 to 1.4, but otherwise keep everything the same. I just copied our patched `vnc_lite.html` into `index.html`, and just split out the JS / CSS. I also changed the title of the page
- prettier with pre-commit has done its thing on the JS, CSS and HTML. This is fine, as I am hoping we can make more changes to the UI after this.
- Modify the `Dockerfile` to setup the conda env first, and then separately install the python package with pip. This makes local development *much* faster.
- Add a `.gitignore` (which this repo was missing), with node_modules included as well as the dist/ directory with the built JS.
- Add a test to make sure the built JS is included in the sdist and bdist
- Add a test to make sure the built container image has the built JS

### CODE REVIEW

**Comment by yuvipanda** [comment]
mmmmmmmmmmm love it 
Screen Shot 2024-02-03 at 1 42 43 AM

**Comment by yuvipanda** [comment]
Once this gets merged, I'll cleanup the JS file a little and implement a version of #7

**Comment by yuvipanda** [comment]
https://github.com/novnc/noVNC/blob/master/docs/API.md there's also a bunch of stuff we can do to make the experience better, like automatic reconnects, sending resize events through, save screenshots, etc.

File: js/index.js [hunk 1]
```diff
@@ -0,0 +1,142 @@
+/**
+ * Derived from https://github.com/novnc/noVNC/blob/master/vnc_lite.html, which was licensed
```
Comment (manics) [inline js/index.js:2]:
```suggestion
 * Derived from https://github.com/novnc/noVNC/blob/v1.4.0/vnc_lite.html, which was licensed
```
I know it's unlikely, but if the licensing on the master branch changes in future including the version will avoid confusion

**Comment by manics** [comment]
The desktop fails to load on mybinder:
`Loading failed for the <script> with source “https://notebooks.gesis.org/binder/jupyter/user/jupyterhub-jupy-e-desktop-proxy-rk6lnz86/desktop/dist/viewer.js”.`

**Comment by yuvipanda** [comment]
@manics ah interesting, it works for me locally in a container. wonder why. I'll dig through and see.

**Comment by yuvipanda** [comment]
I can reproduce the failure on mybinder! Shall test.

**Comment by yuvipanda** [comment]
Interesting, but it works on a JupyterHub!

**Comment by yuvipanda** [comment]
```
(base) jovyan@jupyter-jupyterhub-2djupy-2de-2ddesktop-2dproxy-2d85ah8316:~$ ls /opt/conda/lib/python3.11/site-packages/jupyter_remote_desktop_proxy/static/
clipboard.svg  index.css  index.html
```

Alright, this is something about how the image is being built.

**Comment by yuvipanda** [comment]
@manics it works now!

**Comment by yuvipanda** [comment]
I'm going to add a quick test to make sure that the wheel and sdist have the built files too

**Comment by yuvipanda** [comment]
Cleaning up some JS on top in https://github.com/jupyterhub/jupyter-remote-desktop-proxy/pull/78

**Review by manics — COMMENTED** [review]
Other than the editable install this looks good!

Since this is an application I think we should add a package.lock (and bump it with dependabot) so we can keep track of all the dependencies but that can be done in a follow-up.

File: Dockerfile [hunk 2]
```diff
@@ -24,9 +24,11 @@ RUN apt-get -y -qq update \
...
+COPY --chown=$NB_UID:$NB_GID environment.yml /tmp
+RUN . /opt/conda/bin/activate && \
+    mamba env update --quiet --file /tmp/environment.yml
 
-RUN cd /opt/install && \
-    . /opt/conda/bin/activate && \
-    mamba env update --quiet --file environment.yml
+COPY --chown=$NB_UID:$NB_GID . /opt/install
+RUN . /opt/conda/bin/activate && \
+    pip install -e /opt/install
```
Comment (manics) [inline Dockerfile:34]: Is this editable install required?

File: .gitignore [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+.dmypy.json
+dmypy.json
+
+# Pyre type checker
+.pyre/
+
+# JS libraries
+node_modules/
+# Built JS files
+jupyter_remote_desktop_proxy/static/dist
```
Comment (consideRatio) [inline .gitignore:134]: Couls you put this on top and separated from the other gitihnore stuff with a comment? Knowing what is added beyond whats boilerplate .gitignore stuff is often relevant.

Example at https://github.com/jupyterhub/jupyterhub-python-repo-template/blob/main/.gitignore

File: Dockerfile (hunk 2 above)
Comment (yuvipanda) [reply Dockerfile:34]: Yes, it makes development *much* *much* easier without any real downside.

I'll add a `CONTRIBUTING.md` with more details once the other PRs get handled.

File: .gitignore (hunk 3 above)
Comment (yuvipanda) [reply .gitignore]: @consideRatio done!

**Review by manics — APPROVED** [review]
🎉🎆🥳

**Comment by yuvipanda** [comment]
yay thank you @manics! Take a look at the other PRs too when you get a minute :D When done, I'll make a release.

**Pull request merged**